src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/measure/ActionMeter.kt

```diff
@@ -76,50 +97,80 @@ class ActionMeter(
+        var actionResult: ActionResult = ActionResult.OK
+        var result: T? = null
+
         try {
-            val result = action()
-            output.write(
-                ActionMetric.Builder(
-                    label = key.label,
-                    result = ActionResult.OK,
-                    start = start,
-                    duration = Duration.between(start, clock.instant())
-                )
-                    .virtualUser(virtualUser)
-                    .observation(observation(result))
-                    .drilldown(w3cPerformanceTimeline.record())
-                    .build()
-            )
+            result = action()
             return result
         } catch (e: Exception) {
-            val result = if (e.representsInterrupt()) {
+            actionResult = if (e.representsInterrupt()) {
                 ActionResult.INTERRUPTED
             } else {
                 ActionResult.ERROR
             }
-            output.write(
-                ActionMetric.Builder(
-                    label = key.label,
-                    result = result,
-                    start = start,
-                    duration = Duration.between(start, clock.instant())
-                )
-                    .virtualUser(virtualUser)
+            throw Exception("Action '${key.label}' $actionResult", e)
+        } finally {
+            val metricBuilder = ActionMetric.Builder(
+                label = key.label,
+                result = actionResult,
+                start = start,
+                duration = Duration.between(start, clock.instant())
+            )
+                .virtualUser(virtualUser)
+
+            result?.let { metricBuilder.observation(observation(result)) }
+
+            var metric = metricBuilder.build()
+
+            if (drilldownCondition.test(metric)) {
+                metric = ActionMetric.Builder(metric)
                     .drilldown(w3cPerformanceTimeline.record())
                     .build()
-            )
-            throw Exception("Action '${key.label}' $result", e)
+            }
+            output.write(metric)
         }
     }
 
+    @Deprecated( message = "Use builder instead")
     fun withW3cPerformanceTimeline(
         w3cPerformanceTimeline: W3cPerformanceTimeline
     ): ActionMeter = ActionMeter(
         virtualUser = virtualUser,
         output = output,
         clock = clock,
-        w3cPerformanceTimeline = w3cPerformanceTimeline
+        w3cPerformanceTimeline = w3cPerformanceTimeline,
+        drilldownCondition = Predicate { true }
     )
+
+    class Builder(
+        private val virtualUser: UUID,
+        private val output: ActionMetricOutput,
+        private val clock: Clock
+    ){
+        private var w3cPerformanceTimeline: W3cPerformanceTimeline = DisabledW3cPerformanceTimeline()
+        private var drilldownCondition: Predicate<ActionMetric> = Predicate { true }
+
+        constructor(meter: ActionMeter): this(
```

Comment:
With the `var` avoidance, we might not even need a copying builder here.

Reply:
not sure, if I follow. You mean instead of:
`ActionMeter.Builder(meter)`
invoking it as:
`ActionMeter.Builder(meter.virtualUser, meter.output)` ?
The one before seems more clear, or you mean reusing old builder to build new copy?

Reply:
Oops, you're right, this one is ok. I meant the `ActionMetric` copying builder. Imma comment there